### PR TITLE
[sw/silicon_creator] Make `GlobalMock<Mock>` a base class to support multiple types of mocks for the same module

### PR DESCRIPTION
**Pull request by alphan**
Context: Shutdown module performs a large number of OTP reads (100+) to configure the alert handler and shutdown_unittest models the data in OTP using a struct and a nice mock to keep tests manageable. Since there are other tests that use a strict mock for OTP accesses and our framework does not support multiple types of mocks for the same mask rom module, shutdown_unittest uses a separate definition of `MockOtp` which is not ideal.

This PR addresses this issue by making `GlobalMock<Mock>` defined in `mask_rom_test.h` a base class that mock classes should derive from (CRTP). This allows us to define different types of mocks using type aliases that can be used in tests as we do today. Having `GlobalMock<Mock>` as the base class also allows us to enforce that there is at most one instance of `Mock` by construction (regardless of its strictness or niceness) without using additional data structures.

This approach does not require any changes in existing tests (except for removing MockOtp from shutdown_unittest) but we need to touch existing mocks to use the new class.

### CODE REVIEW

**Comment by alphan** [comment]
@cfrantz and @mcy, PTAL. I would like to hear your thoughts before updating existing mocks.

**Comment by alphan** [comment]
Updated rest of the mocks.

File: sw/device/silicon_creator/lib/base/mock_abs_mmio.h [hunk 1]
```diff
@@ -14,18 +14,16 @@ namespace internal {
 /**
  * Mock class for abs_mmio.c.
  */
-class MockAbsMmio {
+class MockAbsMmio : public GlobalMock<MockAbsMmio> {
```
Comment (cfrantz) [inline sw/device/silicon_creator/lib/base/mock_abs_mmio.h:17]: I had no idea you could do this.  Somehow, this feels like a circular reference: `MockAbsMmio` derives from `GloblaMock` which is parameterized with `MockAbsMmio`.

Its perfectly clear in the `GlobalMock` implementation what is happening... just feels a bit strange to see a derived class mention itself in the base class.

**Review by cfrantz — APPROVED** [review]

File: sw/device/silicon_creator/lib/base/mock_abs_mmio.h (hunk 1 above)
Comment (alphan) [reply sw/device/silicon_creator/lib/base/mock_abs_mmio.h:17]: Thanks for the review @cfrantz! This pattern is called the "curiously recurring template pattern" (CRTP). You can read more about it [here](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern).

**Comment by alphan** [comment]
Merging to enable other PRs.

**Pull request merged**